src/v/cluster/topic_table.cc

```diff
@@ -40,26 +40,33 @@ topic_table::transform_topics(Func&& f) const {
 topic_table::topic_metadata::topic_metadata(
   topic_configuration_assignment c, model::revision_id rid) noexcept
   : configuration(std::move(c))
-  , _id_or_topic(rid) {}
+  , _source_topic(std::nullopt)
+  , _revision(rid) {}
 
 topic_table::topic_metadata::topic_metadata(
-  topic_configuration_assignment c, model::topic st) noexcept
+  topic_configuration_assignment c,
+  model::revision_id rid,
+  model::topic st) noexcept
   : configuration(std::move(c))
-  , _id_or_topic(std::move(st)) {}
+  , _source_topic(st)
+  , _revision(rid) {}
 
 bool topic_table::topic_metadata::is_topic_replicable() const {
-    return std::holds_alternative<model::revision_id>(_id_or_topic);
+    return _source_topic.has_value() == false;
 }
+
 model::revision_id topic_table::topic_metadata::get_revision() const {
     vassert(
       is_topic_replicable(), "Query for revision_id on a non-replicable topic");
-    return std::get<model::revision_id>(_id_or_topic);
+    return _revision;
 }
```

Comment:
the revision is stored in the non-repl case but we can't access it because of the assert?